File: backend/app/api/v1/endpoints/feedback.py
"""API endpoints for user feedback submissions."""
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.postgres import get_db
from app.models.user import User
from app.models.feedback import Feedback
from app.schemas.feedback import FeedbackCreate, FeedbackResponse

router = APIRouter()


@router.post("/", response_model=FeedbackResponse, status_code=status.HTTP_201_CREATED)
async def submit_feedback(
    payload: FeedbackCreate,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> FeedbackResponse:
    """Persist a feedback message together with a metadata snapshot.

    The client supplies free-form metadata in ``client_meta`` (viewport,
    screen size, dpr, route). User identity is taken from the JWT —
    never from the request body — so a malicious client cannot spoof it.
    """
    record = Feedback(
        user_id=current_user.id,
        user_email=current_user.email,
        message=payload.message,
        url=payload.url,
        user_agent=payload.user_agent,
        client_meta=payload.client_meta or {},
    )
    db.add(record)
    await db.commit()
    await db.refresh(record)
    return FeedbackResponse(id=record.id, created_at=record.created_at)

File: frontend/src/services/feedback.ts
import api from '../lib/api'

export interface FeedbackClientMeta {
  viewport?: { width: number; height: number }
  screen?: { width: number; height: number }
  dpr?: number
  language?: string
  platform?: string
  timezone?: string
}

export interface FeedbackPayload {
  message: string
  url?: string
  user_agent?: string
  client_meta?: FeedbackClientMeta
}

export interface FeedbackResponse {
  id: string
  created_at: string
}

export const feedbackService = {
  submit: async (payload: FeedbackPayload): Promise<FeedbackResponse> => {
    const response = await api.post('/feedback/', payload)
    return response.data
  },
}
